File: backend/app/schemas/resource.py
"""
Resource and ResourceRequest Pydantic schemas.
Handles validation for resource management and request system.
"""

from pydantic import BaseModel, Field, HttpUrl
from typing import Optional
from datetime import datetime
from uuid import UUID
from enum import Enum


class ResourceType(str, Enum):
    """Resource type enumeration."""
    PDF = "pdf"
    VIDEO = "video"
    LINK = "link"
    NOTES = "notes"


class RequestStatus(str, Enum):
    """Resource request status enumeration."""
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


# Resource Schemas

class ResourceCreate(BaseModel):
    """Schema for creating a new resource."""
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    type: ResourceType
    url: Optional[str] = None  # For video/link types
    content: Optional[str] = None  # For notes type
    class_id: UUID
    
    class Config:
        from_attributes = True


class ResourceUpdate(BaseModel):
    """Schema for updating a resource."""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    url: Optional[str] = None
    content: Optional[str] = None
    
    class Config:
        from_attributes = True


class TeacherInfo(BaseModel):
    """Basic teacher information."""
    id: UUID
    first_name: str
    last_name: str
    email: str
    
    class Config:
        from_attributes = True


class ResourceResponse(BaseModel):
    """Schema for resource response."""
    id: UUID
    title: str
    description: Optional[str]
    type: ResourceType
    file_path: Optional[str]
    url: Optional[str]
    content: Optional[str]
    teacher_id: UUID
    class_id: UUID
    created_at: datetime
    updated_at: datetime
    teacher: Optional[TeacherInfo] = None
    
    class Config:
        from_attributes = True


# Resource Request Schemas

class ResourceRequestCreate(BaseModel):
    """Schema for creating a resource request."""
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    type: ResourceType
    class_id: UUID
    
    class Config:
        from_attributes = True


class ResourceRequestAction(BaseModel):
    """Schema for teacher action on resource request."""
    teacher_response: Optional[str] = None
    
    class Config:
        from_attributes = True


class StudentInfo(BaseModel):
    """Basic student information."""
    id: UUID
    first_name: str
    last_name: str
    email: str
    
    class Config:
        from_attributes = True


class ResourceRequestResponse(BaseModel):
    """Schema for resource request response."""
    id: UUID
    title: str
    description: Optional[str]
    type: ResourceType
    student_id: UUID
    class_id: UUID
    status: RequestStatus
    teacher_response: Optional[str]
    approved_resource_id: Optional[UUID]
    created_at: datetime
    updated_at: datetime
    student: Optional[StudentInfo] = None
    
    class Config:
        from_attributes = True

File: frontend/src/types/resource.ts
/**
 * Resource and ResourceRequest TypeScript interfaces
 */

export const ResourceType = {
    PDF: 'pdf',
    VIDEO: 'video',
    LINK: 'link',
    NOTES: 'notes',
} as const;

export type ResourceType = typeof ResourceType[keyof typeof ResourceType];

export const RequestStatus = {
    PENDING: 'pending',
    APPROVED: 'approved',
    REJECTED: 'rejected',
} as const;

export type RequestStatus = typeof RequestStatus[keyof typeof RequestStatus];

export interface TeacherInfo {
    id: string;
    first_name: string;
    last_name: string;
    email: string;
}

export interface StudentInfo {
    id: string;
    first_name: string;
    last_name: string;
    email: string;
}

export interface Resource {
    id: string;
    title: string;
    description?: string;
    type: ResourceType;
    file_path?: string;
    url?: string;
    content?: string;
    teacher_id: string;
    class_id: string;
    created_at: string;
    updated_at: string;
    teacher?: TeacherInfo;
}

export interface ResourceRequest {
    id: string;
    title: string;
    description?: string;
    type: ResourceType;
    student_id: string;
    class_id: string;
    status: RequestStatus;
    teacher_response?: string;
    approved_resource_id?: string;
    created_at: string;
    updated_at: string;
    student?: StudentInfo;
}

export interface ResourceCreateData {
    title: string;
    description?: string;
    type: ResourceType;
    class_id: string;
    url?: string;
    content?: string;
    file?: File;
}

export interface ResourceUpdateData {
    title?: string;
    description?: string;
    url?: string;
    content?: string;
}

export interface ResourceRequestCreateData {
    title: string;
    description?: string;
    type: ResourceType;
    class_id: string;
}

export interface ResourceRequestActionData {
    teacher_response?: string;
}
